module/zfs/dmu_traverse.c

```diff
@@ -308,16 +314,51 @@ traverse_visitbp(traverse_data_t *td, const dnode_phys_t *dnp,
 
 		czb = kmem_alloc(sizeof (zbookmark_phys_t), KM_SLEEP);
 
+		/*
+		 * recursively visitbp() blocks below this.
+		 * Indirect (Non L0) Block of size 128k could contain, 1024
+		 * block pointers of 128 bytes. In case of full traverse OR
+		 * incremental traverse, where all blocks were modified, it
+		 * could traverse large number of blocks pointed by indirect.
+		 * Prefetching all blocks in go, could result into large number
+		 * of async reads queued on vdev queue. So, account for prefetch
+		 * issued for blocks pointed by indirect and limit max prefetch
+		 * in one go to zfs_traverse_indirect_prefetch_limit.
+		 */
+		prefetchtriggeridx = 0;
+		prefetchidx = 0;
+		prefetched = 0;
+		prefetchlimit = MIN(epb, zfs_traverse_indirect_prefetch_limit);
+		ASSERT(prefetchtriggeridx <= prefetchidx);
 		for (i = 0; i < epb; i++) {
-			SET_BOOKMARK(czb, zb->zb_objset, zb->zb_object,
-			    zb->zb_level - 1,
-			    zb->zb_blkid * epb + i);
-			traverse_prefetch_metadata(td,
-			    &((blkptr_t *)buf->b_data)[i], czb);
-		}
-
-		/* recursively visitbp() blocks below this */
-		for (i = 0; i < epb; i++) {
+			if (prefetchlimit && i == prefetchtriggeridx &&
+			    prefetchidx < epb) {
+				ASSERT(prefetchtriggeridx <= prefetchidx);
+				prefetched = 0;
+				for (j = prefetchidx; j < epb &&
+				    prefetched < prefetchlimit; j++) {
+					SET_BOOKMARK(czb, zb->zb_objset,
+					    zb->zb_object, zb->zb_level - 1,
+					    zb->zb_blkid * epb + j);
+					if (traverse_prefetch_metadata(td,
+					    &((blkptr_t *)buf->b_data)[j], czb)
+					    == B_TRUE) {
+						prefetched++;
+						if (prefetched ==
+						    MAX((prefetchlimit >> 1),
+						    1))
```

Comment:
This is excessively obfuscated, plus I think it still won't work if prefetchlimit is set to 1, since prefetchtriggeridx will be set to 0 first time, which will never match any upcoming i value, so this code will never be called again.

Reply:
Seems to be mis-reading the code here.  when prefertchlimit is set to 1,  then MAX((prefetchllmit >> 1), 1), would be 1 and in every iteration of "for loop", prefetchtriggeridx would be moved by 1. Like, at prefetchidx 0, it prefetched 1 block and prefetchtriggeridx is set to 1 (next prefetchidx). Then at prefetchidx 1, it again prefetched 1 block and prefetchtriggeridx is set to 2 (next prefetchidx) and so on... So, both prefetchidx and prefetchtriggeridx is moving ahead by 1.

Reply:
As I read it, on the first iteration i, prefetchidx and j are all 0, so when the condition match prefetchtriggeridx is assigned j, which is still 0.  But on next iterations i will be 1 and up, which will never be equal to prefetchtriggeridx of 0.

Reply:
My mistake. You are correct. Prefetch trigger index should always be >= i. Logic of prefetch trigger breaks, when prefetch trigger index is < i;  which is possible when prefetchlimit is set to 1 and all contiguous blocks getting prefetched;
I would make a change to address it. Thanks for pointing it.

Reply:
I still don't like this piece of code with added condition below.  The added condition I'd move up, something like `ptidx = MAX(pidx, i+1);`, but may be something better.  I'd replace ( >> 1) with just /2. It should be trivial enough for compiler to guess, while this code is pretty tangled already.  May be using unsigned variables could just help it a bit, since signs make no sense here.  Also prefetchlimit of 1 makes no sense now, since it prefetches exactly the same block as read below.  I am thinking whether fixing it could somehow simplify all this logic.

Reply:
Ok.  I would move added condition up, as suggested. And change variables type to unsigned. 
Replacing ">>1" with just /2, is not clear, can you elaborate bit more to clarify what exactly you mean ?
As prefetch might not be much effective for very low value of prefetchllmit, so may be, can define some minimum value for prefetch limit, say,
     #define ZFS_TRAVERSE_INDIRECT_PREFETCH_LIMIT_MIN 10
      pefetchlimit = MAX(zfs_traverse_indirect_prefetch_limit, ZFS_TRAVERSE_INDIRECT_PREFETCH_LIMIT_MIN);

Reply:
I mean that compilers are good enough to replace /2 with shifts by themselves. No need to obfuscate the code with excessive optimizations.  In questionable cases I verify the code disassembling it in debugger.

Pathologically useless there I think only pefetchlimit of 1. 2 and up I think should work better.  I was thinking whether some nice algorithm improvement could be found for it, so that it would prefetch the next block instead of the current one.  We could generally not prefetch the first block, since it is almost immediately demand read after that, but then it may happen that demand I/O get to the disk after number of prefetch reads, increasing latency.  Or we could prefetch the first block with SYNC_READ priority and the next ones as ASYNC_READ as they are now.  And then pefetchlimit of 1 could mean "one ahead of the current".

Reply:
Thanks for clarifying about /2 and shift.
Starting with "ptdix = 0; pidx = 1;", should do prefetch of next block instead of current block.